support plt and act

nuomi2021 opened the pull request:
replace https://github.com/ffvvc/FFmpeg/pull/224

@QSXW  this is the refined version. 
Please review and help upstream it after the SEI patch is merged.

nuomi2021 committed:
add github workflow

frankplow committed:
Add fuzz regression test

frankplow committed:
Fix fuzz test path

nuomi2021 committed:
workflow: remove yasm build, since it's disabled by ffmpeg upstream

nuomi2021 committed:
add fuzz failed test

nuomi2021 committed:
workflow: add dvb 40 frames tests

nuomi2021 committed:
workflow: use fuzz flag for fuzz clips

QSXW commented on libavcodec/vvc/dsp_template.c:
```diff
@@ -91,6 +91,25 @@ static void FUNC(transform_bdpcm)(int *coeffs, const int width, const int height
+
+    const int size = width * height;
+    for (int i = 0; i < size; i++) {
+        const int y0 = av_clip(y[i],  min, max);
```
There is an extra space before min.

nuomi2021 replied:
fixed and change av_clip to av_clip_intp2

QSXW commented on libavcodec/vvc/ctu.c:
```diff
@@ -1805,9 +1816,291 @@ static int inter_data(VVCLocalContext *lc)
     return ret;
 }
 
+static TransformUnit* palette_add_tu(VVCLocalContext *lc, const int start, const int end, const VVCTreeType tree_type)
```
The `TransformUnit*` need to be `TransformUnit *`

nuomi2021 replied:
We use this style for all functions that return a pointer, such as add_tu. This means the function should return a TransformUnit pointer.

QSXW commented on libavcodec/vvc/ctu.c:
```diff
@@ -1805,9 +1816,291 @@ static int inter_data(VVCLocalContext *lc)
+}
+
+static void palette_update_predictor(VVCLocalContext *lc, const bool local_dual_tree, int start, int end,
+    bool *predictor_reused, const int predictor_size)
```
Can we use `bool`?

nuomi2021 replied:
yes. c99 has bool

QSXW commented on libavcodec/vvc/ctu.c:
```diff
@@ -1805,9 +1816,291 @@ static int inter_data(VVCLocalContext *lc)
+        Palette *plt = cu->plt + c;
+        int i = cu->plt[start].size;;
+
+        //copy unused predictors to the end of plt
```
better to use `// copy unused predictors to the end of plt`

nuomi2021 replied:
fixed

QSXW commented on libavcodec/vvc/ctu.c:
```diff
@@ -1805,9 +1816,291 @@ static int inter_data(VVCLocalContext *lc)
+        const int hs       = sps->hshift[c];
+        const int vs       = sps->vshift[c];
+        uint8_t *u8        = (uint8_t*)tb->coeffs;
+        uint16_t *u16      = (uint16_t*)tb->coeffs;
```
`(uint16_t*)` to `(uint16_t *)

nuomi2021 replied:
fixed

nuomi2021 commented:
@QSXW  thank you for the review. All fixed.

QSXW commented on Changelog:
```diff
@@ -52,7 +52,8 @@ version 7.1:
   constrains
 - FFV1 parser
-
+- VVC decoder supports all content of SCC (Screen Content Coding):
```
I added this message to the part of version 7.1... Could you help move it below the version <next>: And  the signed off email is wrong, should be outlook.com

nuomi2021 replied:
Oh, you can change this before you send it to upstream for review.

cus committed:
Revert "fftools/textformat/avtextformat: Make close functions return void"

This reverts commit 7684243fbe6e84fecb4a039195d5fda8a006a2a4 and
a888975a3c25760027cd59932f5c1ad04368db8b.

cus committed:
fftools/textformat: correctly propagate uninit error codes

This allows catching IO errors occuring at file close.

Signed-off-by: Marton Balint <cus@passwd.hu>

cus committed:
tests/fate/probe: add test for pcm misdetected as mp3 in wav

Signed-off-by: Marton Balint <cus@passwd.hu>

cus committed:
tests/fate/probe: add test for dts in wav

Signed-off-by: Marton Balint <cus@passwd.hu>

QSXW committed:
avcodec/vvc/cabac: add 9.3.3.5 k-th order Exp - Golomb binarization process

Signed-off-by: Wu Jianhua <toqsxw@outlook.com>

QSXW committed:
avcodec/vvc/cabac: add 9.3.3.7 Fixed-length binarization process

Signed-off-by: Wu Jianhua <toqsxw@outlook.com>

QSXW committed:
avcodec/vvc/cabac: add palette support

Signed-off-by: Wu Jianhua <toqsxw@outlook.com>

QSXW committed:
avcodec/vvc: add VVC_MAX_NUM_PALETTE_PREDICTOR_SIZE

Signed-off-by: Wu Jianhua <toqsxw@outlook.com>

QSXW committed:
avcodec/vvc/ctu: refact out ff_vvc_channel_range

Signed-off-by: Wu Jianhua <toqsxw@outlook.com>

QSXW committed:
avcodec/vvc: refact out ep_init and ep_init_wpp

Signed-off-by: Wu Jianhua <toqsxw@outlook.com>

QSXW committed:
avcodec/vvc: refact, save pf and ciip_flag in ff_vvc_set_intra_mvf

Signed-off-by: Wu Jianhua <toqsxw@outlook.com>

QSXW committed:
avcodec/vvc/ctu: refact out intra_data

Signed-off-by: Wu Jianhua <toqsxw@outlook.com>

QSXW committed:
avcodec/vvc/intra: add ff_vvc_palette_derive_scale

Signed-off-by: Wu Jianhua <toqsxw@outlook.com>

QSXW committed:
avcodec/vvc/ctu: add palette support

Signed-off-by: Wu Jianhua <toqsxw@outlook.com>

QSXW committed:
avcodec/vvc/filter: skip deblocking filter for palette

Signed-off-by: Wu Jianhua <toqsxw@outlook.com>

QSXW committed:
avcodec/vvc/intra: add palette coding decoder

Introduction at https://ieeexplore.ieee.org/document/9408666

passed files:
    10b422_G_Sony_5.bit
    10b422_H_Sony_5.bit
    10b422_I_Sony_5.bit
    10b422_J_Sony_5.bit
    10b422_K_Sony_5.bit
    10b422_L_Sony_5.bit
    8b422_G_Sony_5.bit
    8b422_H_Sony_5.bit
    8b422_I_Sony_5.bit
    8b422_J_Sony_5.bit
    8b422_K_Sony_5.bit
    8b422_L_Sony_5.bit
    8b444_A_Kwai_2.bit
    8b444_B_Kwai_2.bit
    PALETTE_A_Alibaba_2.bit
    PALETTE_B_Alibaba_2.bit
    PALETTE_C_Alibaba_2.bit
    PALETTE_D_Alibaba_2.bit
    PALETTE_E_Alibaba_2.bit

Signed-off-by: Wu Jianhua <toqsxw@outlook.com>

QSXW committed:
avcodec/vvc/cabac: add ff_vvc_cu_act_enabled_flag

Signed-off-by: Wu Jianhua <toqsxw@outlook.com>

QSXW committed:
avcodec/vvc/ctu: read act_enabled_flag for adaptive color transform

Signed-off-by: Wu Jianhua <toqsxw@outlook.com>

QSXW committed:
avcodec/vvc/ctu: fix derive_chroma_intra_pred_mode

See 8.4.3 Derivation process for chroma intra prediction mode

Signed-off-by: Wu Jianhua <toqsxw@outlook.com>

QSXW committed:
avcodec/vvc/dsp: update the interface of pred_residual_joint for joint cbcr residual functionality

Signed-off-by: Wu Jianhua <toqsxw@outlook.com>

QSXW committed:
avcodec/vvc/dsp: add adaptive_color_transform

See 8.7.4.6 Residual modification process for blocks using colour space conversion

Signed-off-by: Wu Jianhua <toqsxw@outlook.com>

QSXW committed:
avcodec/vvc/intra: fix scaling process for transform coefficients

See 8.7.3 Scaling process for transform coefficients

Signed-off-by: Wu Jianhua <toqsxw@outlook.com>

QSXW committed:
avcodec/vvc/intra: refact, predict jcbcr to tb->coeffs

prepare for adaptive color transform

Signed-off-by: Wu Jianhua <toqsxw@outlook.com>

QSXW committed:
avcodec/vvc/intra: make lmcs_scale_chroma inplace

prepare for adaptive color transform

Signed-off-by: Wu Jianhua <toqsxw@outlook.com>

QSXW committed:
avcodec/vvc/intra: refact out lmcs_scale_chroma and add_residual

prepare for adaptive color transform

Signed-off-by: Wu Jianhua <toqsxw@outlook.com>

QSXW committed:
avcodec/vvc: add adaptive color transform support

passed files:
    ACT_A_Kwai_3.bit
    ACT_B_Kwai_3.bit

Signed-off-by: Wu Jianhua <toqsxw@outlook.com>

QSXW committed:
Changelog: VVC supports all content of SCC

Signed-off-by: Wu Jianhua <toqsxw@outlook.com>

nuomi2021 commented:
merged as a65d028fb15dc01ae4974f413e489b891c3964be
Thank you, Jianhua and all
